Bug:  The TooEarly  error during replication

### What happened?

I've set up a replication between two buckets on the same host. A few times a day I see errors in the logs:

```
 [ERROR] -- reductstore/src/replication/replication.rs:108 Failed to read record: [TooEarly] Record with timestamp 1707325657415000 is still being written
```

### Version

1.8.1

### OS

Linux

### Snippet to repeat the bug

1. Provision a ReductStore instance with the following environment variables:
```
"RS_API_TOKEN=REDUCTCOOL2",
    "RS_CERT_PATH=/etc/letsencrypt/live/play.reduct.store/fullchain.pem",
    "RS_CERT_KEY_PATH=/etc/letsencrypt/live/play.reduct.store/privkey.pem",
    "RS_LOG_LEVEL=INFO",

    "RS_BUCKET_1_NAME=datasets",
    "RS_TOKEN_1_NAME=dataset-read",
    "RS_TOKEN_1_VALUE=dataset-read-eab13e4f5f2df1e64363806443eea7ba83406ce701d49378d2f54cfbf02850f5",
    "RS_TOKEN_1_READ=datasets",

    "RS_BUCKET_2_NAME=demo",
    "RS_BUCKET_2_QUOTA_TYPE=FIFO",
    "RS_BUCKET_2_QUOTA_SIZE=10Gb",
    "RS_TOKEN_2_NAME=demo-rw",
    "RS_TOKEN_2_VALUE=reductstore",
    "RS_TOKEN_2_READ=demo,datasets",
    "RS_TOKEN_2_WRITE=demo",

    "RS_BUCKET_3_NAME=stress_test",
    "RS_BUCKET_3_QUOTA_TYPE=FIFO",
    "RS_BUCKET_3_QUOTA_SIZE=20Gb",
    "RS_TOKEN_3_NAME=stress_test-rw",
    "RS_TOKEN_3_READ=stress_test",
    "RS_TOKEN_3_WRITE=stress_test",
    "RS_TOKEN_3_VALUE=rw-b6837ce9a8b9ec849845b96454d1edbd",
    "RS_TOKEN_3_FULL_ACCESS=false",


    "RS_REPLICATION_1_NAME=stress_test",
    "RS_REPLICATION_1_SRC_BUCKET=stress_test",
    "RS_REPLICATION_1_DST_BUCKET=demo",
    "RS_REPLICATION_1_DST_HOST=https://play.reduct.store",
    "RS_REPLICATION_1_DST_TOKEN=reductstore",
    "RS_REPLICATION_1_INCLUDE_size=medium",
```

2. Use https://github.com/reductstore/data-checker to send data to the `stress_test` bucket periodically

### Logs

```Shell
2024-01-28 23:02:00.067 (63840) [INFO] -- reductstore/src/main.rs:24 ReductStore 1.8.1 [e4e743d at 2024-01-28T22:57:22Z]
2024-01-28 23:02:00.067 (63840) [INFO] -- reductstore/src/main.rs:37 License: BUSL-1.1 [https://github.com/reductstore/reductstore/blob/v1.8.1/LICENSE]
2024-01-28 23:02:00.127 (63840) [INFO] -- reductstore/src/main.rs:44 Configuration:
        RS_LOG_LEVEL = INFO
        RS_HOST = 0.0.0.0 (default)
        RS_PORT = 8383 (default)
        RS_API_BASE_PATH = / (default)
        RS_DATA_PATH = /data (default)
        RS_API_TOKEN = ***********
        RS_CERT_PATH = *****************************************************
        RS_CERT_KEY_PATH = ***************************************************
        RS_BUCKET_1_NAME = datasets
        RS_BUCKET_2_NAME = demo
        RS_BUCKET_3_NAME = stress_test
        RS_BUCKET_2_QUOTA_TYPE = FIFO
        RS_BUCKET_2_QUOTA_SIZE = 10.0 GB
        RS_BUCKET_3_QUOTA_TYPE = FIFO
        RS_BUCKET_3_QUOTA_SIZE = 20.0 GB
        RS_TOKEN_1_NAME = dataset-read
        RS_TOKEN_2_NAME = demo-rw
        RS_TOKEN_3_NAME = stress_test-rw
        RS_TOKEN_1_VALUE = *****************************************************************************
        RS_TOKEN_3_VALUE = ***********************************
        RS_TOKEN_2_VALUE = ***********
        RS_TOKEN_1_READ = datasets
        RS_TOKEN_3_READ = stress_test
        RS_TOKEN_3_WRITE = stress_test
        RS_TOKEN_3_FULL_ACCESS = false
        RS_TOKEN_2_READ = demo,datasets
        RS_TOKEN_2_WRITE = demo
        RS_REPLICATION_1_NAME = stress_test
        RS_REPLICATION_1_SRC_BUCKET = stress_test
        RS_REPLICATION_1_DST_BUCKET = demo
        RS_REPLICATION_1_DST_HOST = https://play.reduct.store
        RS_REPLICATION_1_DST_TOKEN = reductstore
        RS_REPLICATION_1_INCLUDE_size = medium

2024-01-28 23:02:14.823 (63840) [INFO] -- reductstore/src/storage/storage.rs:43 Load 3 buckets
2024-01-28 23:02:14.823 (63840) [INFO] -- reductstore/src/cfg.rs:138 Provisioned bucket 'datasets' with: BucketSettings { quota_type: Some(NONE), quota_size: Some(0), max_block_size: Some(64000000), max_block_records: Some(256) }
2024-01-28 23:02:14.823 (63840) [INFO] -- reductstore/src/cfg.rs:138 Provisioned bucket 'stress_test' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(20000000000), max_block_size: Some(64000000), max_block_records: Some(256) }
2024-01-28 23:02:14.824 (63840) [INFO] -- reductstore/src/cfg.rs:138 Provisioned bucket 'demo' with: BucketSettings { quota_type: Some(FIFO), quota_size: Some(10000000000), max_block_size: Some(64000000), max_block_records: Some(256) }
2024-01-28 23:02:14.833 (63840) [INFO] -- reductstore/src/cfg.rs:106 Provisioned token 'dataset-read' with Some(Permissions { full_access: false, read: ["datasets"], write: [""] })
2024-01-28 23:02:14.833 (63840) [INFO] -- reductstore/src/cfg.rs:106 Provisioned token 'stress_test-rw' with Some(Permissions { full_access: false, read: ["stress_test"], write: ["stress_test"] })
2024-01-28 23:02:14.833 (63840) [INFO] -- reductstore/src/cfg.rs:106 Provisioned token 'demo-rw' with Some(Permissions { full_access: false, read: ["demo", "datasets"], write: ["demo"] })
2024-01-28 23:02:14.833 (63840) [INFO] -- reductstore/src/cfg.rs:342 Load Web Console
2024-01-28 23:02:15.598 (63840) [INFO] -- reductstore/src/cfg.rs:168 Provisioned replication 'stress_test' with ReplicationSettings { src_bucket: "stress_test", dst_bucket: "demo", dst_host: "https://play.reduct.store/", dst_token: "***", entries: [], include: {"size": "medium"}, exclude: {} }
2024-01-28 23:02:15.598 (63840) [INFO] -- reductstore/src/main.rs:54 Run HTTP reductstore on https://0.0.0.0:8383/
2024-02-07 18:55:00.562 (97152) [ERROR] -- reductstore/src/replication/replication.rs:108 Failed to read record: [TooEarly] Record with timestamp 1707332100334000 is still being written
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -10,6 +10,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - RS-162: Fix locked block by read/write operation, [PR-398](https://github.com/reductstore/reductstore/pull/398)
+- RS-155: Wait for incomplete records during replication, [PR-399](https://github.com/reductstore/reductstore/pull/399)
 
 ## [1.8.1] - 2024-01-28
 

--- a/reductstore/src/replication/replication.rs
+++ b/reductstore/src/replication/replication.rs
@@ -93,10 +93,8 @@ impl Replication {
                             debug!("Replicating transaction {}/{:?}", entry_name, transaction);
 
                             let read_record_from_storage = async {
-                                let mut record = None;
-                                let mut attempts = 3;
-                                while record.is_none() && attempts > 0 {
-                                    attempts -= 1;
+                                let mut atempts = 3;
+                                loop {
                                     let read_record = async {
                                         thr_storage
                                             .read()
@@ -105,32 +103,35 @@ impl Replication {
                                             .begin_read(&entry_name, *transaction.timestamp())
                                             .await
                                     };
-                                    let rec = match read_record.await {
-                                        Ok(record) => record,
+                                    let record = read_record.await;
+                                    match record {
                                         Err(ReductError {
                                             status: ErrorCode::TooEarly,
                                             ..
                                         }) => {
                                             debug!("Transaction is too early, retrying later");
                                             sleep(Duration::from_millis(10)).await;
-                                            continue;
+                                            atempts -= 1;
                                         }
-                                        Err(err) => {
-                                            log.write().await.pop_front().await.unwrap();
-                                            error!("Failed to read record: {}", err);
-                                            thr_hourly_diagnostics.write().await.count(Err(err));
-                                            break;
+
+                                        _ => {
+                                            atempts = 0;
                                         }
-                                    };
-                                    record = Some(rec);
+                                    }
+
+                                    if atempts == 0 {
+                                        break record;
+                                    }
                                 }
-                                record
                             };
 
                             let record_to_sync = match read_record_from_storage.await {
-                                Some(record) => record,
-                                None => {
-                                    continue;
+                                Ok(record) => record,
+                                Err(err) => {
+                                    log.write().await.pop_front().await.unwrap();
+                                    error!("Failed to read record: {}", err);
+                                    thr_hourly_diagnostics.write().await.count(Err(err));
+                                    break;
                                 }
                             };
 
@@ -472,6 +473,62 @@ mod tests {
                 }
             }
         )
+    }
+
+    #[rstest]
+    #[tokio::test]
+    async fn test_replication_too_early(
+        mut remote_bucket: MockRmBucket,
+        mut notification: TransactionNotification,
+    ) {
+        remote_bucket.expect_write_record().returning(|_, _| Ok(()));
+        remote_bucket.expect_is_active().return_const(true);
+        let replication = build_replication(remote_bucket).await;
+
+        let _tx = replication
+            .storage
+            .write()
+            .await
+            .get_mut_bucket("src")
+            .unwrap()
+            .write_record("test", 20, 4, "".to_string(), Labels::new())
+            .await
+            .unwrap();
+        notification.event = Transaction::WriteRecord(20);
+        replication.notify(notification).await.unwrap();
+        assert!(
+            !transaction_log_is_empty(&replication).await,
+            "We keep the transaction in the log to sync later"
+        );
+        assert_eq!(
+            replication.info().await,
+            ReplicationInfo {
+                name: "test".to_string(),
+                is_active: true,
+                is_provisioned: false,
+                pending_records: 1,
+            }
+        );
+        assert_eq!(replication.diagnostics().await, Diagnostics::default(),);
+
+        sleep(Duration::from_millis(50)).await;
+        assert_eq!(
+            replication.diagnostics().await,
+            Diagnostics {
+                hourly: DiagnosticsItem {
+                    ok: 0,
+                    errored: 60,
+                    errors: HashMap::from_iter(vec![(
+                        425,
+                        DiagnosticsError {
+                            count: 1,
+                            last_message: "Record with timestamp 20 is still being written"
+                                .to_string(),
+                        }
+                    )]),
+                }
+            }
+        );
     }
 
     #[fixture]

--- a/reductstore/src/replication/replication.rs
+++ b/reductstore/src/replication/replication.rs
@@ -8,7 +8,7 @@ use crate::replication::TransactionNotification;
 use crate::storage::storage::Storage;
 
 use log::{debug, error, info};
-use reduct_base::error::ReductError;
+use reduct_base::error::{ErrorCode, ReductError};
 
 use crate::replication::diagnostics::DiagnosticsCounter;
 use reduct_base::msg::diagnostics::Diagnostics;
@@ -92,27 +92,50 @@ impl Replication {
 
                             debug!("Replicating transaction {}/{:?}", entry_name, transaction);
 
-                            let get_record = async {
-                                thr_storage
-                                    .read()
-                                    .await
-                                    .get_bucket(&config.src_bucket)?
-                                    .begin_read(&entry_name, *transaction.timestamp())
-                                    .await
+                            let read_record_from_storage = async {
+                                let mut record = None;
+                                let mut attempts = 3;
+                                while record.is_none() && attempts > 0 {
+                                    attempts -= 1;
+                                    let read_record = async {
+                                        thr_storage
+                                            .read()
+                                            .await
+                                            .get_bucket(&config.src_bucket)?
+                                            .begin_read(&entry_name, *transaction.timestamp())
+                                            .await
+                                    };
+                                    let rec = match read_record.await {
+                                        Ok(record) => record,
+                                        Err(ReductError {
+                                            status: ErrorCode::TooEarly,
+                                            ..
+                                        }) => {
+                                            debug!("Transaction is too early, retrying later");
+                                            sleep(Duration::from_millis(10)).await;
+                                            continue;
+                                        }
+                                        Err(err) => {
+                                            log.write().await.pop_front().await.unwrap();
+                                            error!("Failed to read record: {}", err);
+                                            thr_hourly_diagnostics.write().await.count(Err(err));
+                                            break;
+                                        }
+                                    };
+                                    record = Some(rec);
+                                }
+                                record
                             };
 
-                            let read_record = match get_record.await {
-                                Ok(record) => record,
-                                Err(err) => {
-                                    log.write().await.pop_front().await.unwrap();
-                                    error!("Failed to read record: {}", err);
-                                    thr_hourly_diagnostics.write().await.count(Err(err));
-                                    break;
+                            let record_to_sync = match read_record_from_storage.await {
+                                Some(record) => record,
+                                None => {
+                                    continue;
                                 }
                             };
 
                             let mut bucket = thr_bucket.write().await;
-                            match bucket.write_record(entry_name, read_record).await {
+                            match bucket.write_record(entry_name, record_to_sync).await {
                                 Ok(_) => {
                                     if bucket.is_active() {
                                         thr_hourly_diagnostics.write().await.count(Ok(()));

--- a/reductstore/src/storage/block_manager.rs
+++ b/reductstore/src/storage/block_manager.rs
@@ -547,7 +547,6 @@ pub async fn spawn_write_task(
 #[cfg(test)]
 mod tests {
     use super::*;
-    use bytes::Buf;
     use rstest::{fixture, rstest};
     use tempfile::tempdir;
     use tokio::io::AsyncWriteExt;

